### ISSUE TITLE
STY: comply with the documented line length limit

### ISSUE THREAD

**Issue by lucascolley** [scipy/scipy#19479]
### Is your feature request related to a problem? Please describe.

Following gh-19461, it would help reviewers if CI could catch when lines exceed 88 characters.

We can't enable this in CI right now, since linting is done on whole files at once, so CI would fail on PRs due to violations in untouched lines. This is undesirable since the docs say that:

> It is recommended to leave existing style issues alone unless they exist in lines of code you are already modifying. This practice ensures that the codebase is gradually cleaned up without dedicating precious review time to style-only cleanups.

### Describe the solution you'd like.

To reach a point where we can enable the check in CI, we will need to deal with the ~2500 current violations (the number is so large mainly due to `scipy/constants/_codata.py` - adding a few `noqa`s there reduces the number to about 1000). I propose that we do so by:

1. Reducing line length or adding line breaks where appropriate.
2. Otherwise, adding `# noqa: E501` where we allow lines to exceed the limit.

I think that this should be done one submodule at a time. Perhaps a candidate for `'good first issue'` (cleaning up one submodule)?

### Describe alternatives you've considered.

1. Do nothing. Feel free to close this issue if anyone feels strongly for this option. This would require no extra immediate effort and may be preferable if the time to review PRs for this just doesn't seem worth it. This option will, however, inevitably lead to more violations of the limit creeping in as it is unrealistic to expect reviewers to check every line of every PR for violations manually.
2. Have one big PR to deal with them all at once. It might be easier to keep it isolated to one PR although the diff could get scarily large.

### Additional Information

- Not every violation would have to be dealt with - some of them occur in vendored git submodules, and others could be left alone if it is unlikely that any PRs will be touching that file any time soon. I would estimate that there are about 500 separate places which need a change, since many of the violations are the same structure across multiple lines or in files that won't be modified regularly.
- `# noqa: E501` is already used 18 times in SciPy.

**Comment by mdhaber** [comment]
> # noqa: E501 is already used 18 times in SciPy.

There are other uses of generic `# noqa` for this purpose. I am guilty.

> I would estimate that there are about 500 separate places which need a change, since many of the violations are the same structure across multiple lines or in files that won't be modified regularly.

This sounds managable. I'd be willing to help.

**Comment by j-bowhay** [comment]
I am also +1 on this. I agree breaking this up into submodules sounds like a good idea then we can enable `E501` module by module as the work is completed

**Comment by lucascolley** [comment]
> This sounds managable. I'd be willing to help.

> I am also +1 on this. I agree breaking this up into submodules sounds like a good idea then we can enable E501 module by module as the work is completed

Great! I’ll leave this open for a while but if there are no objections then I’m happy to get the ball rolling with a few PRs.

If we want to enable module by module then sounds like the first step is to switch from the blanket ignore to per-module ignores in `lint.toml`. I’m happy to set up a tracker issue if that’s wanted.

**Comment by lucascolley** [comment]
Proof of concept for the submodule-by-submodule approach at gh-19489.

**Comment by mdhaber** [comment]
Just to clarify, it doesn't looke like this means "vendored git submodule" as in the top post, but rather SciPy "subpackage"?

So you are making fixes to a few subpackages so that you don't have to except 501 for these subpackages. That's fine, since it doesn't look like much is needed, but can we see a commit where the linting *fails* in at least one of these packages so that we know that the line limit check is really working when it's supposed to?

Can we start reviewing that PR?

**Comment by lucascolley** [comment]
> Just to clarify, it doesn't looke like this means "vendored git submodule" as in the top post, but rather SciPy "subpackage"?

What I meant was that I assume we don't need to fix errors in `highs` or `array_api_compat`. Either that or they will just stay ignored in the linter. I propose that we fix or `noqa` the errors in SciPy subpackages so that the linter can check them automatically in future.

> On second thought, can we take those out to keep things organized, and really go through one subpackage at a time?

Yeah, I split them all into separate commits because I imagined that we might want separate PRs. I put them all together into the proof of concept since they are all of the sub packages with only a few changes needed - the rest are more substantial and definitely deserve PRs of their own.

> Also, can we see a commit where the linting fails when the 501 exceptions are removed so that we know that the line limit check is really working when it's supposed to?

Sure, I can split the submodule commits into two in that case.

> Can we start reviewing that PR?

Fine with me! Just didn't want to waste anyone's time if a different direction is taken.

**Comment by mdhaber** [comment]
> What I meant was that I assume we don't need to fix errors in highs or array_api_compat. Either that or they will just stay ignored in the linter. I propose that we fix or noqa the errors in SciPy subpackages so that the linter can check them automatically in future.

Right. That's fine. Git submodules are out of scope. Really, I am suggesting we use "subpackage" here when it's not immediately obvious that "submodule" doesn't mean "vendored git submodule".

> I put them all together into the proof of concept since they are all of the sub packages with only a few changes needed.

OK, that's fine. But if they require *any* changes, I'd suggest taking them out of the real PR. You can open them up one by one as follow-up PRs, even though they're easy.

**Comment by lucascolley** [comment]
> I put them all together into the proof of concept since they are all of the sub packages with only a few changes needed.
OK, that's fine. But if they require any changes, I'd suggest taking them out of this first PR. You can open them up one by one as follow-up PRs, even though they're easy.

Okay, I'll sort that out and open the follow-ups now.

**Comment by mdhaber** [comment]
If you're willing, I'd just create the branches, but hold off on opening all the PRs. I think it is better for lessons learned on the first to be applied to the second, etc. Once we have a few under our belt we can consider opening a few at a time, but I think things will get done just as fast if we go sequentially.

**Comment by rgommers** [comment]
My comment on this now that ~5 PRs have been merged: with this kind of topic, once it's decided that we go ahead and do it (which is the case here now), I'd appreciate thinking about how to get it out of the way as quickly as possible. There is no need for this to be 20 or more PRs, with multiple commits and CI runs for each. Rather, try bunching in larger chunks and not trying to review / second-guess every single decision on individual lines. If it's overall a significant improvement and CI is green, just merge it.

My meta concern here is that between linting, tweaking deprecation messages, bike-sheddy topics like the `>>> import numpy as np` line, etc., there is a very high a level of activity and notifications on these topics, and it detracts from more important things (new features, bug fixes, docs, performance improvements, and more structural maintenance related topics like dealing with unmaintainable Fortran code and build & packaging topics).

tl;dr thank you for dealing with this, and let's try to close the book on this as efficiently as possible.

**Comment by lucascolley** [comment]
> Rather, try bunching in larger chunks and not trying to review / second-guess every single decision on individual lines. If it's overall a significant improvement and CI is green, just merge it.

I'll submit one final PR for now for `constants`, then perhaps someone can try to take the rest on all in one go.

**Comment by mdhaber** [comment]
It was separate PRs by my request, so that's on me. But now that things are rolling, it's fine with me to combine.

**Comment by lucascolley** [comment]
With all of the linked PRs now merged, we are left with these ignores in `lint.toml` (to deal with in a minimal number of separate PRs):

```
# E501 (line length) ignores
"benchmarks/benchmarks/**" = ["E501"]
"scipy/_lib/**" = ["E501"]
"scipy/cluster/**" = ["E501"]
"scipy/constants/_codata.py" = ["E501"]
"scipy/integrate/**" = ["E501"]
"scipy/interpolate/**" = ["E501"]
"scipy/io/**" = ["E501"]
"scipy/linalg/**" = ["E501"]
"scipy/ndimage/**" = ["E501"]
"scipy/optimize/**" = ["E501"]
"scipy/signal/**" = ["E501"]
"scipy/sparse/**" = ["E501"]
"scipy/spatial/**" = ["E501"]
"scipy/special/**" = ["E501"]
"scipy/stats/**" = ["E501"]
"tools/**" = ["E501"]
```

**Comment by j-bowhay** [comment]
I should be able to do the remaining in one PR

**Comment by lucascolley** [comment]
> I should be able to do the remaining in one PR

I would hold off until gh-19516 is in, but go for it after that!

**Comment by lucascolley** [comment]
@j-bowhay have you started work on this? I'm happy to take it on if you want, your choice.

**Comment by j-bowhay** [comment]
Go for it, I slightly over estimated how much free time I have!
